onnxruntime/core/graph/graph_utils.cc

```diff
@@ -598,16 +622,22 @@ void AddNodeInput(Node& target, int target_input_idx, NodeArg& new_input) {
   target.MutableInputArgsCount()[target_input_idx] = 1;
 }
 
-void FinalizeNodeFusion(Graph& graph, Node& first_node, Node& second_node, Node* replacement_node) {
-  graph_utils::RemoveNodeOutputEdges(graph, first_node);
-  graph_utils::MoveAllNodeOutputs(graph, second_node, replacement_node ? *replacement_node : first_node);
+void FinalizeNodeFusion(Graph& graph, Node& first_node, Node& second_node) {
+  // move the outputs from second_node to first_node
+  RemoveNodeOutputEdges(graph, first_node);
+  MoveAllNodeOutputs(graph, second_node, first_node);
 
   // second node now has no output edges and can be removed
   graph.RemoveNode(second_node.Index());
+}
+
+void FinalizeNodeFusion(Graph& graph, const std::vector<std::reference_wrapper<Node>>& nodes, Node& replacement_node) {
```

Comment:
Do we need check that output edges of nodes is either pointing to one of these nodes, or already in replacement_node output edges?

For example, graph edges:
A --> E
B --> D
We fused {A, B} to C, and we might miss the edge for E:
C->D
? ->E

Reply:
That wouldn't be a valid fusion. You can't remove A if it has multiple output edges as it's original output goes away. These checks are done in the fusion implementations though so we assume the correct thing has been done there and it isn't worth re-doing them here.

Reply:
It's fine. It is worth some comments about assumptions. There is an extra assumption: input edges are in first node. In some cases, input edges might exist in several nodes before fusion (like Attention) so they cannot use this function.

Reply:
Will add a comment